iosApp/iosApp/Song Playing/SongPlayingViewModel.swift

```diff
@@ -0,0 +1,122 @@
+import Foundation
+import Combine
+
+protocol SongInfoUseCase {
+	func execute() -> SongPlayingViewModel.Song
+}
+
+protocol MusicPlayerUseCase {
+	var playbackTime: AnyPublisher<Int, Never> { get }
+}
+
+class SongPlayingViewModel: ObservableObject {
+	typealias Seconds = Int
+	
+	private var cancellables = Set<AnyCancellable>()
+	
+	private let songInfoUseCase: SongInfoUseCase
+	private let musicPlayerUseCase: MusicPlayerUseCase
+	
+	@Published var song: Song
+	@Published var lyricsOffsetPercentage: Float = 0
+	@Published var playbackTime: String = SongPlayingViewModel.playbackTimePlaceholder
+	@Published var playbackState: PlaybackState = .paused
+	@Published var showTipping: Bool = false
+	
+	let supportArtistPrompt: String = NSLocalizedString("LIKE_THE_SONG_SUPPORT_THE_ARTIST", comment: "")
+	
+	init(songInfoUseCase: SongInfoUseCase, musicPlayerUseCase: MusicPlayerUseCase) {
+		self.songInfoUseCase = songInfoUseCase
+		self.musicPlayerUseCase = musicPlayerUseCase
+		
+		song = songInfoUseCase.execute()
+		
+		musicPlayerUseCase.playbackTime
+			.map(\.playbackTimeString)
+			.sink { [weak self] in self?.playbackTime = $0 }
+			.store(in: &cancellables)
+	}
+	
+	func supportArtistTapped() {
+		showTipping = true
+	}
+	
+	func previousSongTapped() {
+		
+	}
+	
+	func nextSongTapped() {
+		
+	}
+	
+	func repeatTapped() {
+		
+	}
+	
+	func shuffleTapped() {
+		
+	}
+	
+	func playTapped() {
+		playbackState = .playing
+	}
+	
+	func pauseTapped() {
+		playbackState = .paused
+	}
+	
+	func airplayTapped() {
+		
+	}
+	
+	func shareTapped() {
```

Comment:
Do you think it makes sense to create an enum for the different actions and adding a single function to handle them? 

`func onAction(action: SongPlayingAction)` 

and then just have a switch to dispatch the actions internally? This would keep the api for the viewmodel a bit leaner imo

Reply:
I generally avoid wrapping function names in an enum, never cared for the pattern just because it seems like you're just swapping method names for enum cases and giving it a namespace, which sounds like what a protocol should do.  If you just want all the actions grouped together so it's easier to grok and more organized, I could make a ViewModelActionHandler protocol, and move these methods to an extension?